src/test/java/teammates/test/cases/logic/TaskQueuesLogicTest.java

```diff
@@ -0,0 +1,91 @@
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.Test;
+
+import teammates.common.util.Const.TaskQueue;
+import teammates.common.util.ThreadHelper;
+import teammates.logic.core.TaskQueuesLogic;
+import teammates.test.cases.BaseComponentTestCase;
+import teammates.test.driver.GaeSimulation;
+
+import com.google.appengine.api.taskqueue.dev.LocalTaskQueueCallback;
+import com.google.appengine.api.urlfetch.URLFetchServicePb.URLFetchRequest;
+
+/**
+ * SUT: {@link TaskQueuesLogic}.
+ */
+public class TaskQueuesLogicTest extends BaseComponentTestCase {
+    
+    private static final TaskQueuesLogic taskQueuesLogic = new TaskQueuesLogic();
+    
+    @Override
+    @BeforeTest
+    public void testSetUp() {
+        gaeSimulation = GaeSimulation.inst();
+        gaeSimulation.setupWithTaskQueueCallbackClass(MockTaskQueueCallback.class);
+    }
+    
+    @SuppressWarnings("serial")
+    public static class MockTaskQueueCallback implements LocalTaskQueueCallback {
+        
+        private static int taskCount;
+        private static final int TASK_QUEUE_RESPONSE_OK = 200;
+        
+        @Override
+        public int execute(URLFetchRequest arg0) {
+            taskCount++;
+            return TASK_QUEUE_RESPONSE_OK;
+        }
+        
+        @Override
+        public void initialize(Map<String, String> arg0) {
+            resetTaskCount();
+        }
+        
+        private static void resetTaskCount() {
+            taskCount = 0;
+        }
+        
+    }
+    
+    @Test
+    public void allTests() {
+        
+        int waitTimeForQueueInMs = 1000; // the buffer time for the task to be queued
+        
+        assertEquals(0, MockTaskQueueCallback.taskCount);
+        
+        ______TS("add tasks for immediate queueing");
+        
+        taskQueuesLogic.addTask(TaskQueue.SEND_EMAIL_QUEUE_NAME, "/workerUrl", new HashMap<String, String>());
+        ThreadHelper.waitFor(waitTimeForQueueInMs);
+        assertEquals(1, MockTaskQueueCallback.taskCount);
+        
+        // add another task from different queue
+        
+        taskQueuesLogic.addTaskMultisetParam(TaskQueue.ADMIN_SEND_EMAIL_QUEUE_NAME, "/workerUrl",
+                                             new HashMap<String, String[]>());
+        ThreadHelper.waitFor(waitTimeForQueueInMs);
+        assertEquals(2, MockTaskQueueCallback.taskCount);
+        
+        MockTaskQueueCallback.resetTaskCount();
+        assertEquals(0, MockTaskQueueCallback.taskCount);
+        
+        ______TS("add task for delayed queueing");
+        
+        taskQueuesLogic.addDeferredTask(TaskQueue.SEND_EMAIL_QUEUE_NAME, "/workerUrl",
+                                        new HashMap<String, String>(), 2000);
+        ThreadHelper.waitFor(waitTimeForQueueInMs + 250);
+        assertEquals(0, MockTaskQueueCallback.taskCount); // task is not queued yet
+        ThreadHelper.waitFor(500);
```

Comment:
Hmmm this is not meant to be `1000 / 2`, it's just an amount such that it has not reached the defer time of 2000 (plus waiting time of 1000) yet. It could be other amounts like 400, 600, etc.